src/ColorPicker/Halogen/Layout.purs

```diff
@@ -0,0 +1,459 @@
+
+type InputTextValue = { value ∷ String, isValid ∷ Boolean }
+
+type PositionUpdate = { x ∷ Number, y ∷ Number } → Dynamic Color
+
+type Dynamic s = LazyColor → s
+
+type Classes = Array H.ClassName
+
+type PreNumConf = { prefix ∷ String, title ∷ String, placeholder ∷ String, range ∷ Range Number }
+
+type LazyColor =
+  { color ∷ Color
+  , hsl ∷ Lazy RecordHSLA
+  , hsv ∷ Lazy RecordHSVA
+  , rgb ∷ Lazy RecordRGBA
+  }
+
+type InputProps c =
+  { root ∷ c
+  , label ∷ c
+  , elem ∷ c
+  , elemInvalid ∷ c
+  }
+
+
+data PickerComponent
+  = NumberComponentSpec
+    { update ∷ Number → Dynamic (Maybe Color)
+    , read ∷ Dynamic Number
+    , props ∷ Num.Props Number
+    , view ∷ NumberComponentView
+    }
+  | TextComponentSpec
+    { fromString ∷ String → Maybe Color
+    , view ∷ TextComponentView
+    }
+  | DragComponentSpec
+    { update ∷ PositionUpdate
+    , view ∷ DragComponentView
+    }
+  | ActionComponentSpec ActionComponentView
+
+type NumberComponentView =
+  ∀ p i
+  . { color ∷ LazyColor
+    , input ∷ HH.HTML p i
+    }
+  → Array (HH.HTML p i)
+
+type TextComponentView =
+  ∀ p i
+  . { color ∷ LazyColor
+    , value ∷ Maybe InputTextValue
+    , onValueInput ∷ String → i
+    , onBlur ∷ FocusEvent → i
+    }
+  → Array (HH.HTML p i)
+
+type DragComponentView =
+  ∀ p i
+  . { color∷ LazyColor
+    , onMouseDown ∷ MouseEvent → i
+    , onTouchStart ∷ TouchEvent → i
+    }
+  → Array (HH.HTML p i)
+
+type ActionComponentView =
+  ∀ p i
+  . { color ∷ ValueHistory LazyColor
+    , setColor ∷ Color → i
+    , commit ∷ i
+    }
+  → Array (HH.HTML p i)
+
+mapValueHistory ∷ ∀ a b. (a → b) → ValueHistory a → ValueHistory b
+mapValueHistory f { current, old } = { current: f current, old: map f old }
+
+mkLazyColor ∷ Color → LazyColor
+mkLazyColor color =
+  { color
+  , hsl: defer \_ → Color.toHSLA color
+  , hsv: defer \_ → Color.toHSVA color
+  , rgb: defer \_ → Color.toRGBA color
+  }
+
+
+componentPreview ∷ Array H.ClassName → PickerComponent
+componentPreview classes = ActionComponentSpec \{ color , setColor } → pure $
+  HH.div
+    [ HP.classes $ classes
+    , HP.title "Current value"
+    , HCSS.style $ CSS.backgroundColor color.current.color
+    ] []
+
+componentHistory ∷ Int → Array H.ClassName → PickerComponent
+componentHistory historySize classes = ActionComponentSpec \{ color , setColor } →
+  take historySize color.old <#> \c →
+    HH.div
+      [ HP.tabIndex 0
+      , HP.classes $ classes
+      , HP.title "Old value"
+      , HE.onClick $ (\_ → Just $ setColor c.color)
+      , HCSS.style $ CSS.backgroundColor c.color
+      ] []
+
+
+componentSet ∷ Array H.ClassName → PickerComponent
+componentSet classes = ActionComponentSpec \{ color: {current, old}, commit } → pure $
+  HH.button
+    [ HP.classes classes
+    , HE.onClick $ const $ Just commit
+    , HP.disabled $ Just current.color == (map _.color $ head old)
+    ]
+    [ HH.text "Set" ]
+
+componentDragSV ∷
+  { isLight ∷ Array H.ClassName
```

Comment:
Using open rows here could help with the situation where one config is used for multiple functions i.e.
```purescript
componentDragHue :: forall r. { root :: Array H.ClassName | r} -> ...
componentDragSV :: forall r. { root :: Array H.ClassName, isDark :: H.ClassName | r } -> ... 
-- May be used with 
{ root: [], isDark: [], isLight: [] } etc
```
BTW, `isDark` looks like boolean flag not an array